### PR TITLE
CVE-2017-14633: Don't allow for more than 256 channels

### PR DESCRIPTION
**Pull request by agx**
Otherwise

 for(i=0;i<vi->channels;i++){
      /* the encoder setup assumes that all the modes used by any
         specific bitrate tweaking use the same floor */
      int submap=info->chmuxlist[i];

overreads later in mapping0_forward since chmuxlist is a fixed array of
256 elements max.

This fixes https://gitlab.xiph.org/xiph/vorbis/issues/2329 for me but I don't know much about vorbis so I might be wrong.

### CODE REVIEW

File: lib/info.c [hunk 1]
```diff
@@ -588,7 +588,7 @@ int vorbis_analysis_headerout(vorbis_dsp_state *v,
   oggpack_buffer opb;
   private_state *b=v->backend_state;
 
-  if(!b||vi->channels<=0){
+  if(!b||vi->channels<=0||vi->channels>256){
```
Comment (rillian) [inline lib/info.c]: This should be `> 255` I think?

Do you have a way for vi->channels to fail this check? The value is initialized from an one-byte read in the info packet data, so memory corruption is the only way I see for a value >255 to occur.

**Comment by agx** [comment]
@rillian According to backend.h we can have 256 elements:
    int   chmuxlist[256]
and the check is for < vi-> channels so 0..255 is o.k. 

I see the bug during encoding:

which uses vorbis_encode_setup_setting and there's no such check:

      vi->channels=channels;

I can move the check over there if that's the better place.

**Comment by agx** [comment]
Ping?

**Comment by agx** [comment]
The current code has a bug where if channels <=0 or >= 256 the code would call .oggpack_writeclear on a not yet initialized pack. If added a commit to fix this too.

**Comment by Tsunamical** [comment]
It would be nice to see this merged and then 1.3.6 shortly after

**Comment by tdaede** [comment]
Rebased and merged as a79ec216cd119069c68b8f3542c6a425a74ab993 and c1c2831fc7306d5fbd7bc800324efd12b28d327f.

A TODO is to move the channel # check earlier, as passing 256 in still won't produce a valid bitstream.

**Pull request closed without merging**